From f2f39845769df446fe32a4338ab7a450e0c4b3d3 Mon Sep 17 00:00:00 2001
From: Zaikuo Wang <zaikuo.wang@intel.com>
Date: Mon, 30 Mar 2015 12:13:57 +0800
Subject: [PATCH 336/338] atomisp2: Fix compat parameters copy stack corruption

Fix a stack corruption when copying the parameters from 32
to 64 bits.

Change-Id: I35dbb10107f63098821d04fe5045a38e5169a3e2
Tracked-On: https://jira01.devtools.intel.com/browse/IMINAN-27747
Signed-off-by: Zaikuo Wang <zaikuo.wang@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/48710
Approver: Bin B Han <bin.b.han@intel.com>
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Build: IMIN-legacy (SC)
Integrator: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.c       | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.c
index 8e319b2..862d349 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_ioctl32.c
@@ -691,18 +691,19 @@ static int put_atomisp_acc_s_mapped_arg32(struct atomisp_acc_s_mapped_arg *kp,
 static int get_atomisp_parameters32(struct atomisp_parameters *kp,
 					struct atomisp_parameters32 __user *up)
 {
-	unsigned int n = sizeof(struct atomisp_parameters32) /
+	int n = offsetof(struct atomisp_parameters32, output_frame) /
 				sizeof(compat_uptr_t);
 
 	if (!access_ok(VERIFY_READ, up, sizeof(struct atomisp_parameters32)))
 			return -EFAULT;
 
-	while (n-- > 0) {
+	while (n >= 0) {
 		compat_uptr_t *src = (compat_uptr_t *)up + n;
 		uintptr_t *dst = (uintptr_t *)kp + n;
 
 		if (get_user((*dst), src))
 			return -EFAULT;
+		n--;
 	}
 	if (get_user(kp->isp_config_id, &up->isp_config_id) ||
 	    get_user(kp->per_frame_setting, &up->per_frame_setting))
-- 
1.9.1

